### ISSUE TITLE
Improve human readability of json validator output

### ISSUE THREAD

**Issue by pieterlukasse** [opentargets/platform#401]
Upon reviewing and testing https://github.com/opentargets/json_schema/pull/69, I noticed one part which is not really backwards compatible: the real error messages from the validation are lost now due to message truncating. Example:

In previous version, if I would make a mistake like adding property "drugs" instead of "drug":
```
{
    "access_level": "private",
    "disease": {
        "id": "http://www.ebi.ac.uk/efo/EFO_0003829"
    },
    "drugs": {
  ...
```
I would get an error like:
```
2019-01-14 16:16:51,820 - opentargets_validator.validator - ERROR - failed validating line 1 eval 
0.00654602050781 secs with these errors u'drug' is a required property ;  ; Failed validating u'required' 
in schema: ;     {u'$schema': u'http://json-schema.org/draft-07/schema#', ;      u'allOf': [{u'$ref': 
u'base.json'}, ;                 {u'properties': {u'drug': {u'$ref': u'bioentity/drug.json', ;                 
                           u'type': u'object'}, ;                                  u'evidence': {u'properties': {u'drug2clinic': 
{u'$ref': u'evidence/drug/drug2clinic.json', ;                                                                                 u'type': 
u'object'}, ;                                                                u'target2drug': {u'$ref': 
u'evidence/drug/target2drug.json', ;                                                                                 u'type': u'object'}}, 
;
                                                u'required': [u'target2drug', ;                                                              
u'drug2clinic'], ;                                                u'type': u'object'}, ;                                  u'type': {u'enum': 
[u'known_drug'], ;                                            u'type': u'string'}}}], ;      u'description': u'Schema for target-disease evidence where there is drug knowledge', ;      u'properties': {u'access_level': {}, ;                      
u'disease': {}, ;                      u'drug': {}, ;                      u'evidence': {}, ;                      u'literature': {}, ;                      u'sourceID': {}, ;                      u'target': {}, ;                      u'type': {}, ;                      u'unique_association_fields': {}}, ;      u'required': [u'type', ;                    u'access_level', ;                    
u'sourceID', ;                    u'unique_association_fields', ;                    u'target', ;                    u'drug', ;         
           u'disease', ;                    u'evidence'], ;      u'title': u'Drug-based evidence strings', ;      u'type': u'object', ;      u'version': u'1.3.0'} ;  ; On instance: ;     {u'access_level' ; 
...
```
But now I get an error like:
```
2019-01-14 16:13:25,652 - opentargets_validator.validator - ERROR - failed validating line 1 eval 
0.0454699993134 secs with these errors {u'target': {u'target_name': u'Neuronal acetylcholine receptor; 
alpha4/beta2', u'target_type': u'http://identifiers.org/cttv.target/protein_evidence', u'target_class': 
[u'Nicotinic acetylcholine receptor alpha subunit', u'Nicotinic acetylcholine receptor beta subunit'], u'id': 
u'http://identifiers.org/uniprot/P17787', u'activity': 
u'http://identifiers.org/cttv.activity/drug_positive_modulator'}, u'validated_against_schema_version': 
u'1.2.8', u'sourceI...
....  <----continues with the rest of the JSON line and truncates the message because it is too long
            and I never get to see the real cause of the error.
````
Any ideas on how to fix this?

**Comment by afaulconbridge** [comment]
Thanks for flagging this @pieterlukasse . I agree that it would be good to have more human-readable error messages, but I'm not sure what the best way to achieve that is. In the short term, we can remove the truncation on https://github.com/opentargets/validator/blob/6cba8aa1c6c28c48d21c9167503c950ce6f9df99/opentargets_validator/validator.py#L40-L42

If you know of anywhere (webservice, library in any language!) that reports json validation errors in a better way, please let us know. We'd really like to have some examples of how this should be done rather than spending a lot of time sorting it out ourselves.

**Comment by mkarmona** [comment]
@pieterlukasse  @afaulconbridge ,with the previous version of the reported schema errors, if you replace `' ; '` by `\n` then you will get a nicer output of the reported schema error as it was reported by the schema checker package

**Comment by pieterlukasse** [comment]
hi @afaulconbridge, perhaps the solution could be to break up the log into the error parts. If I understand [this page](https://python-jsonschema.readthedocs.io/en/latest/errors/) correctly, it is possible to get the message, context and cause separately. Currently the code is likely just concatenating all at: https://github.com/opentargets/validator/blob/6cba8aa1c6c28c48d21c9167503c950ce6f9df99/opentargets_validator/validator.py#L30

**Comment by pieterlukasse** [comment]
hi @afaulconbridge, any updates on this one? Trying 19.04 now, but my data is not passing validation anymore. The `explanation_str` in `19.04_invalid-evidence-data` seems to say it is not valid according to the schema...and then dumps the **whole** schema...so it is really hard to find out which part is incorrect.

**Comment by pieterlukasse** [comment]
@cmalangone can you please reopen this one? I tested it and it does not work for me.

**Comment by pieterlukasse** [comment]
@afaulconbridge @cmalangone I think the problem could be related to the fact that the pypi package has not yet been updated:

![image](https://user-images.githubusercontent.com/2900303/59033150-4fe69a80-8868-11e9-9fd3-2ab36ad27f51.png)

**Comment by afaulconbridge** [comment]
@pieterlukasse If you can give details of exactly what you did and how it doesn't work, and I can confirm its a problem, then it can be investigated. Master works for me.

In the mean time, I'll mark this as "done" but tagged as fixed since the version containing it isn't release yet.

**Comment by pieterlukasse** [comment]
hi @afaulconbridge , unfortunately it does not work for me. The latest version just prints out the complete JSON item and says "<my complete json here>...} is not valid under any of the given schemas"

**Comment by pieterlukasse** [comment]
@afaulconbridge I found a way to reproduce it. 

1. make a json item that is valid according to 1.5 json_schema for chembl (known_drug). This item will contain `max_phase_for_disease`
2. now validate this item with 1.6 schema

**Comment by pieterlukasse** [comment]
hi @afaulconbridge, any updates on this one?

**Comment by afaulconbridge** [comment]
@pieterlukasse We've released version 0.5.0 which should improve log output, as well as adding the ability to check if the `unique_association_fields` are actually unique.

Could you give it a try and let us know if it helps?

**Comment by afaulconbridge** [comment]
Done, closing.
